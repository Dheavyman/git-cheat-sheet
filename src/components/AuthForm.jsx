import React from 'react';

/**
 * Authentication form component
 *
 * @param {boolean} forLogin - Type of authentication
 * @param {function} handleChange - Handle input change
 * @param {function} handleSubmit - Handle form submission
 * @param {function} handleToggleAuth - Handle toggling forLogin
 * @param {boolean} isLoading - Form submit status
 */
const AuthForm = ({ forLogin, handleChange, handleSubmit,
  handleToggleAuth, isLoading, errorMessage }) => (
  <div className="row">
    <form className="col s12" onSubmit={handleSubmit}>
      <div className="row">
        <div className="input-field col s12">
          <input
            id="username"
            name="username"
            type="text"
            className="validate"
            onChange={handleChange}
            required
          />
          <label htmlFor="username">Username</label>
        </div>
        <div className="input-field col s12">
          <input
            id="password"
            name="password"
            type="password"
            className="validate"
            onChange={handleChange}
            required
          />
          <label htmlFor="password">Password</label>
        </div>
        {errorMessage &&
        <div className="row center-align red-text">{errorMessage}</div>
        }
        <div className="row center-align">
          <button
            type="submit"
            className="waves-effect waves-light btn"
            disabled={isLoading}
          >
            {forLogin ? 'Login' : 'Register'}
          </button>
        </div>
        <div className="row center-align">
          {forLogin
          ? 'Don\'t have an account? '
          : 'Already have an account? '}
          <a
            href="#!"
            className="cursor-pointer blue-text"
            onClick={handleToggleAuth}
            >
              {forLogin ? 'Register' : 'Login'}
          </a>
        </div>
      </div>
    </form>
  </div>
);

export default AuthForm;
