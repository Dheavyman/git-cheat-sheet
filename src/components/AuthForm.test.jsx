import React from 'react';
import TestRenderer from 'react-test-renderer';

import AuthForm from './AuthForm';

describe('AuthForm component', () => {
  let props;

  beforeEach(() => {
    props = {
      forLogin: true,
      handleChange: jest.fn(),
      handleSubmit: jest.fn(),
      handleToggleAuth: jest.fn(),
      isLoading: false,
    };
  });

  it('should render successfully', () => {
    const renderer = TestRenderer.create(<AuthForm {...props} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should render with register button', () => {
    props = {
      ...props,
      forLogin: false,
    };
    const renderer = TestRenderer.create(<AuthForm {...props} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should call handleChange when username input changes', () => {
    const renderer = TestRenderer.create(<AuthForm {...props} />);

    renderer.root.findAllByType('input')[0].props.onChange();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call handleChange when password input changes', () => {
    const renderer = TestRenderer.create(<AuthForm {...props} />);

    renderer.root.findAllByType('input')[1].props.onChange();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call handleToggleAuth when changing form', () => {
    const renderer = TestRenderer.create(<AuthForm {...props} />);

    renderer.root.findAllByType('a')[0].props.onClick();
    expect(props.handleToggleAuth).toHaveBeenCalled();
  });

  it('should disable submit button when loading', () => {
    props = {
      ...props,
      isLoading: true,
    };
    const renderer = TestRenderer.create(<AuthForm {...props} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
