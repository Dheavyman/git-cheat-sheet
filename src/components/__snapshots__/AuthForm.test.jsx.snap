// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm component should disable submit button when loading 1`] = `
<div
  className="row"
>
  <form
    className="col s12"
    onSubmit={[MockFunction]}
  >
    <div
      className="row"
    >
      <div
        className="input-field col s12"
      >
        <input
          className="validate"
          id="username"
          name="username"
          onChange={[MockFunction]}
          required={true}
          type="text"
        />
        <label
          htmlFor="username"
        >
          Username
        </label>
      </div>
      <div
        className="input-field col s12"
      >
        <input
          className="validate"
          id="password"
          name="password"
          onChange={[MockFunction]}
          required={true}
          type="password"
        />
        <label
          htmlFor="password"
        >
          Password
        </label>
      </div>
      
      <div
        className="row center-align"
      >
        <button
          className="waves-effect waves-light btn"
          disabled={true}
          type="submit"
        >
          Login
        </button>
      </div>
      <div
        className="row center-align"
      >
        Don't have an account? 
        <a
          className="cursor-pointer blue-text"
          href="#!"
          onClick={[MockFunction]}
        >
          Register
        </a>
      </div>
    </div>
  </form>
</div>
`;

exports[`AuthForm component should render successfully 1`] = `
<div
  className="row"
>
  <form
    className="col s12"
    onSubmit={[MockFunction]}
  >
    <div
      className="row"
    >
      <div
        className="input-field col s12"
      >
        <input
          className="validate"
          id="username"
          name="username"
          onChange={[MockFunction]}
          required={true}
          type="text"
        />
        <label
          htmlFor="username"
        >
          Username
        </label>
      </div>
      <div
        className="input-field col s12"
      >
        <input
          className="validate"
          id="password"
          name="password"
          onChange={[MockFunction]}
          required={true}
          type="password"
        />
        <label
          htmlFor="password"
        >
          Password
        </label>
      </div>
      
      <div
        className="row center-align"
      >
        <button
          className="waves-effect waves-light btn"
          disabled={false}
          type="submit"
        >
          Login
        </button>
      </div>
      <div
        className="row center-align"
      >
        Don't have an account? 
        <a
          className="cursor-pointer blue-text"
          href="#!"
          onClick={[MockFunction]}
        >
          Register
        </a>
      </div>
    </div>
  </form>
</div>
`;

exports[`AuthForm component should render with register button 1`] = `
<div
  className="row"
>
  <form
    className="col s12"
    onSubmit={[MockFunction]}
  >
    <div
      className="row"
    >
      <div
        className="input-field col s12"
      >
        <input
          className="validate"
          id="username"
          name="username"
          onChange={[MockFunction]}
          required={true}
          type="text"
        />
        <label
          htmlFor="username"
        >
          Username
        </label>
      </div>
      <div
        className="input-field col s12"
      >
        <input
          className="validate"
          id="password"
          name="password"
          onChange={[MockFunction]}
          required={true}
          type="password"
        />
        <label
          htmlFor="password"
        >
          Password
        </label>
      </div>
      
      <div
        className="row center-align"
      >
        <button
          className="waves-effect waves-light btn"
          disabled={false}
          type="submit"
        >
          Register
        </button>
      </div>
      <div
        className="row center-align"
      >
        Already have an account? 
        <a
          className="cursor-pointer blue-text"
          href="#!"
          onClick={[MockFunction]}
        >
          Login
        </a>
      </div>
    </div>
  </form>
</div>
`;
